### PR TITLE
feat(WiFi): Add wifi.configure action

### PR DESCRIPTION
**Pull request by Rapsssito**
# What does this implement/fix?

Added `wifi.configure` action that connects to ab SSID and password (optionally saving it in persistent memory after checking for a successful connection). This allows for more complex setups where the SSID and password might be negotiated dynamically.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#4186

## Test Environment

- [ ] ESP32
- [x] ESP32 IDF
- [ ] ESP8266
- [ ] RP2040
- [ ] BK72xx
- [ ] RTL87xx

## Example entry for `config.yaml`:

```yaml
# Example config.yaml
script:
  - id: change_wifi_sta
    then:
      - wifi.configure:
          ssid: MySSID
          password: password1
          save: true
          on_connect: 
            - logger.log: "Connected to WiFi!"
          on_error:
            - logger.log: "Failed to connect to WiFi!"
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

**Comment by Rapsssito** [comment]
@jesserockz, what do you think? It is quite a small PR and might be very useful for dynamic WiFi connections.

**Comment by jesserockz** [comment]
Not against it, but what would be better would be that the action actually sets the details on the wifi component, makes sure it connects successfully, only then would it save to prefs, otherwise logs an error.

improv does this too:

```c++
  wifi::global_wifi_component->set_sta(sta);
  wifi::global_wifi_component->start_connecting(sta, false);

  ... 
  // Sets state and keeps checking in `loop` if connected before saving.
  ...

  if (wifi::global_wifi_component->is_connected()) {
    wifi::global_wifi_component->save_wifi_sta(this->connecting_sta_.get_ssid(),
                                             this->connecting_sta_.get_password());
  }
```

**Comment by Rapsssito** [comment]
@jesserockz, you are right! I have updated the method so it now checks for a connection before saving it to the device. I had to use `wifi.disable()` and `wifi.enable()` to make it work (the WiFi component does not like to be disconnected from the current network). Let me know what you think!

I will update the documentation once it is approved, to avoid having to change it again.

**Comment by Rapsssito** [comment]
@jesserockz, just pinging you again as a friendly reminder!

**Comment by pzich** [comment]
@jesserockz is this good to merge? Or are there changes that need to be made?

**Comment by jesserockz** [comment]
Will review again after release this week

File: esphome/components/wifi/__init__.py [hunk 1]
```diff
@@ -483,3 +486,36 @@ async def wifi_enable_to_code(config, action_id, template_arg, args):
 @automation.register_action("wifi.disable", WiFiDisableAction, cv.Schema({}))
 async def wifi_disable_to_code(config, action_id, template_arg, args):
     return cg.new_Pvariable(action_id, template_arg)
+
+
+@automation.register_action(
+    "wifi.set_sta",
```
Comment (jesserockz) [inline esphome/components/wifi/__init__.py]: Not sure I like the action name being `set_sta`. This refers to an internal function name and is not exactly user friendly for the yaml side. 
Maybe simply `wifi.connect`?
```suggestion
    "wifi.connect",
```

**Review by jesserockz — CHANGES_REQUESTED** [review]
Further to my requested name change below, I think adding a `save: bool` config to the schema to allow users to not save to prefs (can default to true) would be nice too.

File: esphome/components/wifi/__init__.py (hunk 1 above)
Comment (pzich) [reply esphome/components/wifi/__init__.py]: I feel like `connect` could be confused with just enabling/disabling it. Maybe `update`, `change`, or even just `set`?

Comment (jesserockz) [reply esphome/components/wifi/__init__.py]: This is true, maybe `wifi.configure`?

**Comment by pzich** [comment]
Probably outside of the scope of this PR, but it might be useful if this could optionally take `networks` instead (similar to the wifi config itself) to update/replace the whole set of ssid/pws in the list.

**Comment by Rapsssito** [comment]
@jesserockz, @pzich, I have implemented your suggestions. Let me know what you think!

Regarding multiple networks, sadly I am not familiar with that kind of setup. @pzich, feel free to build on top of my PR!

**Review by jesserockz — APPROVED** [review]

**Pull request merged**